make_prg/from_msa/cluster_sequences.py

```diff
@@ -246,18 +263,20 @@ def kmeans_cluster_seqs_in_interval(
 
     no_clustering = num_clusters == 1 or num_clusters == num_sequences
     if no_clustering:
-        return merge_sequences(
+        clustered_ids = [flatten_list(seq_to_ids.values()) + flatten_list(small_seq_to_ids.values())]
+        merged_sequences = merge_sequences(
             seq_to_ids.keys(), small_seq_to_ids.keys(), first_seq=first_sequence
         )
+        return ClusteringResult(clustered_ids, merged_sequences)
```

Comment:
Just curious here: `merge_sequences` used to produce a `ClusteringResult` object, why do you prefer not to? Could there be an issue with the sequence ids not matching the sequences?

though if i understand correctly you will only ever have both ids and sequences in `ClusteringResult` when there is no clustering, right? So you would only have one big list of all the ids and one big list of all the sequences.

One option is to have a classmethod in `ClusteringResult` called `make_no_clustering`, with argument the ids, and it does what these lines do

again, feel free to ignore, i just want us, when/if we reread this code in 6 months, to understand quickly what it's doing and also to have a clear guarantee that ids and sequences are matched

Reply:
actually come to think of this if the first sequence gets moved by `merged_sequences` function, shouldn't the first sequence id be placed first in `clustered_ids`?

Reply:
> Just curious here: merge_sequences used to produce a ClusteringResult object, why do you prefer not to? Could there be an issue with the sequence ids not matching the sequences?


The main factor for this change is that the previous `merge_sequences()` would return a `ClusteringResult(None, [first_seq] + result)`, i.e. the clustering would be `None`. This is fine for the current code - there is no significant cluster, so let's juts look at the sequences directly. However, in the code of this PR, we explicitly represent the recursion tree, and in some part of this recursion tree, we might have a `MultiClusterNode` (to be seen in next PRs, a node that represent a sub-alignment that will potentially be divided in multiple clusters) that will be split into several `SingleClusterNode`. For the code in the PR, we need to always have the clusters, even if they are not significant, and thus this is why I refactored this function to not return a `ClusteringResult` with no clustering. There is indeed the option to not change this, and keep the return value as `ClusteringResult(None, [first_seq] + result)`, but it is much easier to change this single return value than to change the recursion tree code to accommodate this. The second reason is just for semantics and consistency in the `kmeans_cluster_seqs()` function. We have the `merge_clusters()` functions which merge clusters and return clusters. When I first looked at the `merge_sequences()` function, I thought it would similar: merge sequences and return sequences, but it returned a `ClusteringResult`, which is clusters + sequences. So in `kmeans_cluster_seqs()`, we have a more consistent way to compute the return value: merge clusters, if needed; merge sequences, if needed; and bundle these objects into a `ClusteringResult` object. Please note that I am happy to rollback to the previous code if you think we should, as I think you are the main author of this module.

> though if i understand correctly you will only ever have both ids and sequences in ClusteringResult when there is no clustering, right? So you would only have one big list of all the ids and one big list of all the sequences.

Yes, the recursion tree code always need to know how the clustering was done, even if not significant, to work properly. This is fixable though, we can change it I guess...

> One option is to have a classmethod in ClusteringResult called make_no_clustering, with argument the ids, and it does what these lines do

I'd need further clarifications here, do we keep the old `merge_sequences()`, and add `ClusteringResult.make_no_clustering()`, which is the new `merge_sequences()`?

> again, feel free to ignore, i just want us, when/if we reread this code in 6 months, to understand quickly what it's doing and also to have a clear guarantee that ids and sequences are matched

I don't want to ignore this, this is the crucial part of the review, make sure this looks understandable to both of us and new devs

> actually come to think of this if the first sequence gets moved by merged_sequences function, shouldn't the first sequence id be placed first in clustered_ids?

I think this is yet another bug! Will make a test case for this!

Reply:
So I think this is the key thing about this potential new bug: it might be better to just order the cluster IDs, and then go from cluster IDs to sequences (using the alignment)

Then you don't need to worry about tracking the two in two separate lines- you see what I mean?

Reply:
So do we rollback to what we had before, just having `clustered_ids` instead of `ClusteringResult`?

Reply:
I'm actually not sure what to do- i've looked at previous code and there were already two separate functions `merge_sequences` that places first sequence at beginning and `merge_clusters` that places first cluster ID at beginning

Also `merge_sequences` was only used to produced a `ClusteringResult` object corresponding to no clustering anyway, so this 'bug' here might not be one if you don't actually use cluster IDs anyway

Reply:
True, but will solve the bug anyway, as leaving such a bug is just a time bomb, and will increase technical debt, that we should continuously reduce

Reply:
Fixed this bug and added test cases, please check. Now, in `kmeans_cluster_seqs()`, we always call `merge_clusters()` to merge clusters and to set the id of the first record to the first position of the first cluster, and `merge_sequences()` to merge sequences, if needed, and to set the first sequence in the first position.